### PR TITLE
PyScript Core Documentation

### PR DESCRIPTION
**Pull request by WebReflection**
## Description

This MR would like to at least draft the most comprehensive user-facing documentation around the current *TP* so that it's clear what is offering, what is different, why we made some choice and why we use some terminology, hoping to help the team also to catch up with the current state of `@pyscript/core`.

This MR **also** fixed a few issues in the making!

## Changes

 * created a *docs* folder with all the getting started needed to understand current offer
 * created a file that copy/paste and test integration tests around examples shown in the very same doc
 * fixed a MicroPython typo and one error around `runAsync`, even if not fully implemented
 * fixed a weird *coincident* bug but now all DOM operations are tested to work fine, no workarounds needed!
 * update *coincident* module to latest as it fixes tons of issues around recursive main thread references

## Checklist

-   [x] All tests pass locally
-   [ ] I have updated `docs/changelog.md`
-   [ ] I have created documentation for this(if applicable)

### CODE REVIEW

**Review by tedpatrick — APPROVED** [review]

File: pyscript.core/docs/README.md
Comment (madhur-tandon) [inline pyscript.core/docs/README.md:315]:
```suggestion
However, the native `Worker` utility is incapable of parsing anything that is not *JS* so that we thought that providing a *cross-worker* utility, such as `XWorker` is, would be a great addition to what *core* offers out of the box.
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:184]:
```suggestion
While`<script type="non-standard-type">` works with any browser, it's possible that some of the currently non-standard-types become standard in the future. For this reason, we've chosen to use specific names which we feel are safe (i.e. `<script type="pyodide">` instead of `type="python"`) - users are encouraged to use the same caution when specifying their own type names to avoid conflicts with future standards.
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:196]:
```suggestion
| version   | `<script type="pyodide" version="0.23.2">`    |  Load a specific version of the specified interpreter type. If a numeric version is provided (i.e. `0.23.2`) , that version of the interpreter will be loaded from the interpreter's default CDN. If a fully qualified URL is provided, the interpreter will be loaded from that URL.
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:207]:
```suggestion
If the interpreter has a way to access the [document](https://developer.mozilla.org/en-US/docs/Web/API/Document), the document's `currentScript` attribute will contain a reference to the script tag that is currently executing code (even if its `async` attribute is used). This mimics what the standard [document.currentScript](https://developer.mozilla.org/en-US/docs/Web/API/Document/currentScript) offers already. This property only exists for *synchronous* and blocking scripts that are running, and never interferes with *core* logic or vice-versa.
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:220]: This is where I'd move the description of the `target` property from above - this is really the context where I think it's useful.

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:232]: I think this is the first reference to experimental vs non-experimental interpreters. Do we intend to release any interpreters that are still experimental? If not, I don't think we need to mention them here.

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:201]:
```suggestion
*@pyscript/core* adds the following additional features to all the interpreters it includes:
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:223]: See above on where I think `target` should go.

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:266]:
```suggestion
The prefix for event handlers can be the full name of an interpreter type, as above, or a prefix associated with a custom type alias/extension:
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:275]:
```suggestion
The text within the attribute is executed just like any other inline or external content is in a \<script\> tag with the matching type, and through the same *interpreter*. The only difference is that, within code that is triggered by an event, an temporary global [Event object](https://developer.mozilla.org/en-US/docs/Web/API/Event) called `event` is availble.
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:277]:
```suggestion
This is the same behavior as native Web inline events handlers.

Some key takeaways of this approach:
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:279]: *JG Comment: Did we move away from the idea of passing the event object directly, via the object passed by the JS event listener? Would that solve this issue with asynchronous handlers?*

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:280]: I think unless we're planning to release/encourage users to release experimental interpreters, we can probably cut this.

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:284]:
```suggestion
#### The [type]-env attribute
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:286]:
```suggestion
Just as the `env` attribute on a \<script\> tag specifies a specific instance of an interpreter to use to run code, it is possible to use the `[type]-env` attribute to specify which instance of an interpreter or custom type should be used to run event code:
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:309]:
```suggestion

It is not currently possible for an element to have register event listeners for two different interpreters of the same type using this syntax.
```
I think this is true?

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:315]:
```suggestion
@pyscript/core adds a functionality called `XWorker` to all of the interpreters it offers, which works in each language the way `Worker` does in JavaScript.
```

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [inline pyscript.core/docs/README.md:317]:
```suggestion
In each Interpreter, `XWorker` is a global reference. 
```

**Review by JeffersGlass — COMMENTED** [review]
There's so much good stuff here! I left a bunch of comments on specific sections, mostly tightening up language - feel free to take or leave any/all of those.

I think my big comments on structure would be:

- I think this documentation could use a 'The Structure of a Runtime' section, maybe just before 'How Scripts Work,' describing the basic structure of the objects in `esm/interpreter`. Something like:
  -  "Each interpreter is represented internally by a JavaScript object which has a reference to interpreter's WebAssembly representation, as well as the following methods: `run`, `runAsync`, `setGlobal`... etc etc. I think that's useful context for the attributes that follow
- As noted in the comments, I think you could move the description of what a `target` is down to the section for `document.currentScript`

I didn't get too far into the last XWorker section - I want to make sure I actually understand it fully first 😅

File: pyscript.core/docs/README.md
Comment (madhur-tandon) [reply pyscript.core/docs/README.md:50]: Programming Language most probably

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:52]: That's correct because we provide only 4 interpreters that must be pre-bundle and made available through our artifacts. A custom type builds up on top of that interpreter so I think your change makes sense.

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:58]: We do wtih `display` though in PyScript extend ... but that part hasn't been written yet so maybe you are right.

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:309]: not sure I am following ... any example you have in mind so I can add it to disambiguate / avoid confusion?

**Comment by WebReflection** [comment]
@madhur-tandon , @JeffersGlass thanks a lot for your help and patience and I've flagged this MR out of WIP as I think everything that needs to be known, from a consumer of *core* perspective, is there.

Looking forward to your final thoughts and suggestions but also please be aware I've recently (today) discovered the `sync` feature is not working beside *Wasmoon* case and I need to fix that before merging this, so some code might slip in but in terms of documentation, I hope we're good now.

**Review by tedpatrick — APPROVED** [review]
🚢 it

**Comment by WebReflection** [comment]
@fpliger there's also @ntoll waiting for this to be merged so that a new PR with hints and changes can land ... if that's OK, I am going to merge this into next and I've just published the latest as `0.0.5` to *npm* so everything should be good (in terms of CDN) pretty soon.

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:23]: dozen --> the full standard library and support installing dozens of modules

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:26]: Isn't wasmoon also "experimental"? Or do you mean the interpreter itself is experimental?

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:114]: I find this description a bit confusing... In general, `env` refers to `enviroment`. In this case, it's something like this:

* an environment is an isolated interpreter instance with a specific configuration (packages and files installed into it)
* these environments are identified by "names"
* the `env` attribute on a script allows users to reference a specific environment by name

NOTE:  (Don't want to be picky... just PTSD about communicating the concept of `Enviroments` in the context of Python package managers (`conda`, `pip`, etc..)

File: pyscript.core/docs/README.md
Comment (fpliger) [reply pyscript.core/docs/README.md:130]: 👆🏼

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:133]: It'd be good to add how those seconds add up here (how much can run in parallel and how much is serial)

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:182]: If this is referring to `Custom Elements` and the fact that inline code is broken, I'd be more explicit here...

File: pyscript.core/docs/README.md
Comment (fpliger) [reply pyscript.core/docs/README.md:184]: Probably worth mentioning that we "explicitly take that into account when adding default types in core and any plugin adding aliases or support for new `types` should that that into consideration as well", or something like that

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:192]: Create, if not known yet, a dedicated unique *environment* for the specified `type`. If that environment has already been created, it'll reuse it instead. Please read the [Terminology](#terminology) **env** dedicated details to know more.

File: pyscript.core/docs/README.md
Comment (fpliger) [reply pyscript.core/docs/README.md:207]: +1 on the suggestion, especially specifying that it's the script **tag** as it can be easily misinterpreted with the "script code" itself.

File: pyscript.core/docs/README.md
Comment (fpliger) [reply pyscript.core/docs/README.md:232]: Mmmm... this is the Tech Preview docs. I assume we'll re-work them and change as we work on the next task: "Moving `next` as new core and make the right architecture and high level API changes". The XWorker interface, for instance might change... or we may decide to move the experimental types out of core and make them available as plugins out of core? 

Just want to make sure we are all in agreement that this is not the final/set in stone, architecture/API.

Comment (fpliger) [reply pyscript.core/docs/README.md:232]: I'd also rewrite this in a way to put more emphasis in the `XWorker` API itself rather than the interpreter. Maybe something like:

"
The goal of the`XWorker` API is to enable off-loading heavy operations to a worker, without blocking the main / UI thread (the current page), while also allowing such worker to even reach the `document` or anything else available on the very same main / UI thread.

Note that every (*non experimental*) interpreter will have defined, at the global level, a reference to the `XWorker`, with or without access to the `document`.
"

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:264]: Comment here mostly about future..... I think we'll have to work that muscle a lit bit because (I think!!!!) this will be ambiguous. What happens if you have 2 `micropython` enviroments? In that sense we'd need a way to reference an interpreter by type and/or by name... or something like that. 

Not proposing a solution, just highlighting a possible corner case that we need to consider for the future

File: pyscript.core/docs/README.md
Comment (JeffersGlass) [reply pyscript.core/docs/README.md:309]: For instance, there's no way to have 'click' responders associated to two different envs on the same element, right?

```html
<script type="micropython" env="one">
    def foo():
        print("FOO!")
</script>
<script type="micropython" env="two">
    def bar():
        print("BAR!")
</script>

<!-- 
    There's no way to link this button to functions in both env one and env two, right?
    I.e. Is there a way for a click to run both foo() and bar()? 
    Or for a click to run foo() and a mouseover to run bar()?
-->
<button micropython-env="???" micropython-click="???">Click Me</button>
```

File: pyscript.core/docs/README.md
Comment (fpliger) [reply pyscript.core/docs/README.md:279]: Would be good to also talk about the alternatives or how to prevent this scenario

File: pyscript.core/docs/README.md
Comment (fpliger) [reply pyscript.core/docs/README.md:280]: I agree that, written as is, it can add more confusion. I.e., what should the user do? Are we saying "don't use this feature for experimental interpreters"?

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:341]: Think that using `sync` as variable name here can be more confusing to the end user.... it can become confusing in regards of what's part of the API and what's just the variable name

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:402]: Think the name "Scripts" here is a bit confusing (overloaded term).

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:467]: All of the above is absolutely amazing! I think that adding an example of these would go a looooooong way to help users understand how they can be used (Yes, these are advanced feature but, still, would be great)

File: pyscript.core/docs/README.md
Comment (fpliger) [inline pyscript.core/docs/README.md:470]: missing description here

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:133]: every *env* is a different pyodide, the only removed time with multiple envs is the time to download the module ... let's say 100ms ... if `loadPyodide` takes 1 second to bootstrap, there's no parallel in the same thread, every second will sum up to the amount of envs, the main thread is one, has no parallel work there with WASM modules running in the very same thread.

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:182]: no, it refers to the fact that `<script type="python">` or `<script type="json">` might conflict in the future, there's nothing about *custom elements* in core because no *custom element* in the registry is strictly needed.

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:402]: it's about having custom `<script type="...">` tags ... any suggestion?

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:309]: I think there's no use case for that so I am not sure this is a real-world concern ... *env* is an **advanced** feature and whoever uses it would realize it's never desired to have other envs meshed up with listeners as it completely defeat the purpose of an env ... you bootstrap two different worlds, why would you want to trigger listeners all over the place?

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:26]: wasmoon can do everything Pyodide or MicroPython can do ... why do you think it's experimental?

File: pyscript.core/docs/README.md
Comment (WebReflection) [reply pyscript.core/docs/README.md:114]: `env` here defines a unique environment ... every *env* is a different pyodide or a different MicroPython ... I don't personally know any other terminology to define a runtime/interpreter environment, and *env* plays very well with attributes as it's also very short.

The package manager for the *env* import stuff *only* for that *env*, so which part is not clear or you think should be improved as description? It is, in fact, a unique environment for *micropip* as it won't import anything else in any other pyodide out there that is *not the same env* ... am I misunderstanding your comment or did I really mess up docs in here?

**Pull request merged**
